### ISSUE TITLE
registerBoundHelper-generated helpers do not support use with Handlebars blocks.

### ISSUE THREAD

**Issue by zion03** [emberjs/ember.js#10295]
After update Handlebars to v.2 I got deprecation:
 `DEPRECATION: Usage of Ember.Handlebars.get is deprecated, use a Component or Ember.Handlebars.makeBoundHelper instead.`

But when I rewrite code using BoundHelper I got 
  `registerBoundHelper-generated helpers do not support use with Handlebars blocks.`

helper look like this 
       ```  {{#compare val1 val2}}
         {{else}}
         {{/comapre}}```

**Comment by OrKoN** [comment]
I am interested in a resolution regarding this as well. We have a similar helper {{#equal}} which we need to maintain.

**Comment by mmun** [comment]
@zion03 In 1.10 you should be able be able to use a subexpression:

```hbs
{{#if (compare val1 val2)}} {{else}} {{/if}}
```

**Comment by zion03** [comment]
Thanks, it's works in 1.10+ versions

**Comment by sandstrom** [comment]
For uses where subexpressions cannot be used, how does one register a bound helper with blocks? [in 1.11+]

**Comment by perlun** [comment]
> For uses where subexpressions cannot be used, how does one register a bound helper with blocks? [in 1.11+]

:+1: I would also like to know this...

**Comment by mmun** [comment]
Bound helpers with blocks don't fit well with the old rendering model. It will be possible (and a first class citizen) once Glimmer lands.

**Comment by sandstrom** [comment]
Great! :boat:

**Comment by mmun** [comment]
Take a look at the new `if` helper to get a sense of how it works: https://github.com/emberjs/ember.js/blob/idempotent-rerender/packages/ember-htmlbars/lib/helpers/if_unless.js.

The magic is in the `yield` calls.

**Comment by perlun** [comment]
Somewhat off-topic, but closely related, here is my use case: Any idea how I would transpose this?

```
Handlebars.registerHelper('ew-form', (options) ->
  data = options.data
  fn = options.fn
  view = data.view
  context = view.get('context')

  FormView = Ember.View.extend({
    // implementation removed
  })

  Ember.Handlebars.ViewHelper.helper(this, FormView, options)
)
```

`Ember.Handlebars._registerHelper` is now private, and even if I choose the wrong approach (and try to use it), it still doesn't work like `registerHelper` used to do. `options` is basically empty, `options.data` does not exist and so forth.

The use case is basically: register a helper with a custom view with some even handlers like `willDestroyElement` and some wireup. The helper template is quite simple, just some basic stuff and then a `{{yield}}` block.

**Any** pointers, ideas whatsoever about how to resolve this would be extremely appreciated. Thanks.

**Comment by rwjblue** [comment]
@perlun - The default for these types of block helpers should be to use a component, what caused you do make a helper like this instead of a component (it is not clear from your implementation).

**Comment by perlun** [comment]
@rwjblue, I didn't write the code myself in the first place, but I have now discussed with my colleague who did the technical design on this specific part, and I now have more details about it. The full code looks like this btw:

````
Handlebars.registerHelper('ew-form', (options) ->
  data = options.data
  fn = options.fn
  view = data.view
  context = view.get('context')

  FormView = Ember.View.extend({
    layout: Ember.Handlebars.compile("""
      {{#if view.inline}}
        <form class="form-inline" role="form">
          {{yield}}
        </form>
      {{else}}
        <form {{bind-attr class="view.vertical::form-horizontal"}} role="form">
          <button type="submit" class="ew-hidden-submit"></button>
          {{yield}}
        </form>
      {{/if}}
    """)

    init: ->
      @_super()
      validator = EW.Validator.create()
      validator.set('view', this)
      @set('validator', validator)
      @get('context.validators')?.register(@get('validator'))

    submit: (e) ->
      @$('button.btn-primary')?.click()
      return false

    toString: -> "EmberWidgets.Form"

    willDestroyElement: ->
      @get('context.validators')?.unregister(@get('validator'))
  })

  Ember.Handlebars.ViewHelper.helper(this, FormView, options)
)
```

`context.validators` is a collection that we mix in to all controllers by reopening `Ember.ControllerMixin`.

The chief reason why this was done as a helper rather than a component is because it needs access to the context (i.e. the *controller* of the view in which it is being used), so that it can set up the validator there.

So my questions are now:

1. How can I rephrase the above for Ember 1.10?
1. Will this approach (mucking with the context from a helper) work at all with Ember 1.11+/2.0? (given that you're moving towards the direction of routable components and all)

Huge thanks in advance.

**Comment by sandstrom** [comment]
Hey Per!

I'll try to answer as best as I can.

1. Use a plain component (no handlebars-helper needed).
2. Instead of mucking with the context, use dependency injection[1] to inject the validator[s] into this component (or all components, depending on your need).

*As an aside, the code above will define the class, and compile the template, over and over every time the helper is called. Depending on how often that is, it may needlessly slow down your app.*

[1] http://guides.emberjs.com/v1.10.0/understanding-ember/dependency-injection-and-service-lookup/

*PS. I sent an email to your gmail, is it still in use? DS.*

**Comment by perlun** [comment]
Alexander, thanks for your comment. (yes, I still use my gmail, but haven't replied to your email yet - thanks :smile:).

The problem here is that:

1. You can have multiple `ew-form` instances on the same page.
2. The validator should be scoped to the `ew-form` which it belongs.
3. `context.validators` contains a list of all validators (i.e. forms) in the current context, i.e. in the same view (with the same controller).
4. Child components will be set up so that they get in touch with "their" closest `ew-form` parent so that they can register themselves within that validator. This is used to implement `isValid` and `isInvalid` in this app.

So, I'm not sure that dependency injection isn't really solving the problem here. We specifically want the 1-to-1 connection between one `ew-form` and its validator, so that you can e.g. have a button within the form that is disabled when *that* specific form is invalid. That is crucical.

In other words, the "scope this to anything that is logically DOM children to myself" is the key here.

(Yeah, I know that the dynamic template compilation isn't the smartest thing in the world. :smile: I don't think the sheer number of `ew-form` instances is that big so I don't think the performance impact is *that* bad in this case.)

**Comment by sandstrom** [comment]
Ah, okay. Perhaps you could use a wrapper-component (with block-params) holding the validator?

**Comment by perlun** [comment]
> Ah, okay. Perhaps you could use a wrapper-component (with block-params) holding the validator?

Please elaborate on that. How does it work? My core issue here is also that this is being used by our application framework which we use for a bunch of application, so we really don't want to change the public API. (Ember 1.8 to 1.10 should only be a minor, entirely backwards compatible upgrade, hint hint... :wink:)

**Comment by perlun** [comment]
I understand now what you meant. However, it breaks all existing code + feels very manual. I like our previous approach; you don't have to think about it so much, it's just there doing its job,

@rwjblue - ideas about how to transpose my example to 1.10 or 1.11?

**Comment by sandstrom** [comment]
Below is a bare-bones example. Hopefully it illustrates the main idea.

```coffeescript
EwFormComponent = Ember.Component.extend({  
  validator: null
  
  wiring: (->
    # adjust the validator here, if needed
    @set 'validator', EW.Validator.create()
  ).on('init')
  
  layout: Ember.Handlebars.compile('<form>{{yield validator}}</form>')
})
```

```handlebars
{{#ew-form as |validator|}}
  {{#ew-some-child-component validator=validator}}
    <h1>Hello, I share the validator with my parent form</h1>
  {{/ew-some-child-component}}
{{/ew-form}}
```

**Comment by perlun** [comment]
Thanks @sandstrom, that was how I thought. The problem is that this will get **extremely** awkward in a typical scenario, like so (simplified to illustrate the point):

```
  {{#ew-form as |validator|}}
    <div class="form-compact">
      <div class="row">
        {{ew-autocomplete
          value=selectedFacility
          content=m3lookups.facilities
          validate="required"
          validator=validator}}
        {{ew-autocomplete
          value=selectedWorkCenter
          content=filteredWorkCenters
          validate="required"
          validator=validator}}
        {{ew-daterange
          startValue=startDate
          endValue=endDate}}
        <div class="form-group col-xs-3">
          <div>
            <div class="col-xs-4">
              {{ew-textfield
                inline=true
                validate="number"
                validator=validator
                value=referenceOrderLine}}
            </div>
          </div>
        </div>
        <div class="col-xs-1 pull-right" style="padding-top: 17px; margin-right: 10px;">
          {{ew-button
            action="searchManufacturingOrders"
            disabled=validator.isInvalid
            validator=validator
            text=_Search}}
          </div>
      </div>
    </div>
  {{/ew-form}}
```

I simply don't like it. It adds an extra manual step (4 instances in a simple use case like above) and would force all our developers to remember this (IMHO) stupid "gotcha". It will inevitably lead to people forgetting it at times, which will waste time debugging. Plus, it breaks the very basic [DRY](http://en.wikipedia.org/wiki/Don%27t_repeat_yourself) principle.

I appreciate your helpfulness, so thanks for that, but simply put, that approach won't fit the bill. How can I use a Handlebars helper to accomplish this?

**Comment by sandstrom** [comment]
You could probably use [parentView](http://emberjs.com/api/classes/Ember.Component.html#property_parentView) to access the validator, but it would be more opaque.
